File: pwned.py
#!python3.6

import os
import sys
import hashlib
import collections
from getpass import getpass

SearchResult = collections.namedtuple('SearchResult', ['lineno', 'used', 'tries', 'sha1hash'])
def check_passwd(filename, passwd):
    file_size = os.stat(filename).st_size
    found = False
    tries = 0

    lines = file_size // RECORD_SIZE
    print(f'number of entries: {lines}')
    low = 0
    high = lines
    mid = (low + high) // 2

    with open(filename, 'rb') as psfile:
        while low <= high:
            tries += 1
            mid = (low + high) // 2
            psfile.seek(mid * RECORD_SIZE)
            entry, used = psfile.read(RECORD_SIZE - 1).strip().split(b':')
            entry = entry.decode('ascii')
            used = int(used.decode('ascii'))

            if passwd < entry:
                high = mid - 1
            elif passwd > entry:
                low = mid + 1
            elif passwd == entry:
                return SearchResult(lineno=mid+1, tries=tries, sha1hash=entry, used=used)
    return None


def check_password_and_show_result(filename, password):
    result = check_passwd(filename, password)
    if result is None:
        print("Password was not found")
    else:
        print('Password found on line %d used %d times.\t hash %s tries %d' % (
            result.lineno, result.used, result.sha1hash, result.tries))

if __name__ == "__main__":
    if len(sys.argv) < 2:
        print('Using pwned-passwords-2.0_sorted.txt by default')
        filename = 'pwned-passwords-2.0_sorted.txt'
    else:
      filename = sys.argv[1]
    global RECORD_SIZE
    with open(filename, 'rb') as psfile:
        RECORD_SIZE = len(psfile.readline())
    ps = getpass('enter your password: ')
    passwd_orig = hashlib.sha1(ps.encode('utf-8')).hexdigest().upper()
    passwd_lower = hashlib.sha1(ps.lower().encode('utf-8')).hexdigest().upper()
    passwd_upper = hashlib.sha1(ps.upper().encode('utf-8')).hexdigest().upper()
    print('Original password hash: %s\nYour lower password hash: %s\nupper password hash: %s' % (
        passwd_orig, passwd_lower, passwd_upper))

    print("\nSearch original password")
    result = check_password_and_show_result(filename, passwd_orig)

    print("\nSearch upper case password")
    result = check_password_and_show_result(filename, passwd_upper)

    print("\nSearch lower case password")
    check_password_and_show_result(filename, passwd_lower)

File: pwned.nim
import os
import securehash
import strutils
import unicode

const RECORD_SIZE:int = 42
proc check_passwd(filename, passwd: string): bool =
    let
        file_size = os.getFileSize(filename)
        passwd_file = open(filename, fmRead)
    var
        tries: int64 = 0
        min: int64 = 0
        max: int64 = file_size div RECORD_SIZE
        mid: int64 = (min + max) div 2
        line = newString(40)
    
    while min <= max:
        tries += 1
        mid = (min + max) div 2
        passwd_file.setFilePos(mid * RECORD_SIZE)
        discard passwd_file.readChars(line, 0, 40)
        if passwd > line:
            min = mid + 1
        elif passwd < line:
            max = mid - 1
        elif passwd == line:
            echo("Found line $1\npos $2\ntries $3" % [$line, $mid, $tries])
            return true
    return false



when isMainModule:
    if paramCount() < 1:
        quit "You have to supply password file"
    stdout.write("enter your password: ")
    let
        filename = paramStr(1)
        passwd = stdin.readLine
        passwd_lower = secureHash(unicode.toLower(passwd))
        passwd_upper = secureHash(unicode.toUpper(passwd))
    echo("Your lower password hash: $1\nupper password hash: $2" % [$passwd_lower, $passwd_upper])
    if not check_passwd(filename, $passwd_lower):
        echo("Lower passwd NOT found")
    
    if not check_passwd(filename, $passwd_upper):
        echo("Upper passwd NOT found")
